### PR TITLE
CB-8168 --list support for CLI

### PR DESCRIPTION
**Pull request by muratsu**
Add list support back after the discussion on the dev mail list

### CODE REVIEW

File: doc/run.txt [hunk 1]
```diff
@@ -1,13 +1,14 @@
 Synopsis
 
-    cordova-cli run [BUILD] [PROD] [TARGET] [EXP] [PLATS] [-- POPTS]
+    cordova-cli run [BUILD] [PROD] [TARGET] [EXP] [PLATS] [LIST] [-- POPTS]
```
Comment (jsoref) [inline doc/run.txt]: [LIST] should be before [EXP], and you need to define LIST below (between TARGET and EXP)

Comment (jsoref) [inline doc/run.txt]: Actually, LIST is probably exclusive to BUILD,

so:

Replace [BUILD] with [MODE], and change BUILD to MODE: --list|--nobuild

File: doc/run.txt [hunk 2]
```diff
@@ -20,6 +21,10 @@ Deploys app on specified platform devices / emulators
     --emulator ......................... Deploy to an emulator
     --target ........................... Deploy to a specific target
 
+    --list ............................. Lists out available targets.
```
Comment (jsoref) [inline doc/run.txt]: Drop "out" and drop "."

File: doc/run.txt [hunk 3]
```diff
@@ -1,13 +1,14 @@
...
 
-      BUILD:  --nobuild
-      PROD:   --debug|--release
-      TARGET: --device|--emulator|--target=FOO
-      EXP:    --experimental [EXPERIMENTALFLAGS]
-      PLATS:  PLATFORM [...]
-      POPTS:  platformopts
+      BUILD:       --nobuild
+      PROD:        --debug|--release
+      TARGET:      [DEVICECLASS]|--target=FOO
```
Comment (jsoref) [inline doc/run.txt]: drop the [] around DEVICECLASS.

[] means "optional", but TARGET was already optional above, so it doesn't make sense for DEVICECLASS to be optional too...

File: src/cli.js [hunk 4]
```diff
@@ -206,12 +207,13 @@ function cli(inputArgs) {
         // calling into platform code should be dealing with this based
         // on the parsed args object.
         var downstreamArgs = [];
-        var argNames = [ 'debug', 'release', 'device', 'emulator', 'nobuild'];
+        var argNames = [ 'debug', 'release', 'device', 'emulator', 'nobuild', 'list'];
```
Comment (jsoref) [inline src/cli.js]: Someone should split this array initialization line into multiple lines, one line per arg-name. But that doesn't have to be done in this commit.

(Splitting it means that you don't get blame for the entire line each time you add/remove an arg-name from the list, only for the one you're inserting/removing.)

File: src/cli.js [hunk 5]
```diff
@@ -220,6 +222,11 @@ function cli(inputArgs) {
         }
         opts.options = downstreamArgs.concat(unparsedArgs);
 
+        if (cmd == 'run' && args.list && cordova.raw.targets) {
+            cordova.raw.targets(opts).done();
```
Comment (jsoref) [inline src/cli.js]: Code further down uses .call(null, ...), I'm not quite sure why, but unless there's a good reason not to follow that convention, we should probably follow it.

-- Please ask someone.

**Comment by muratsu** [comment]
@jsoref please take a look

**Pull request closed without merging**